src/core/server/saved_objects/routes/migrate.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IRouter } from '../../http';
+import { IKibanaMigrator } from '../migrations';
+
+export const registerMigrateRoute = (
+  router: IRouter,
+  migratorPromise: Promise<IKibanaMigrator>
+) => {
+  router.post(
+    {
+      path: '/api/saved_objects/_migrate',
+      validate: false,
+    },
+    async (context, req, res) => {
+      const migrator = await migratorPromise;
```

Comment:
Could we log a warning if this endpoint is used in production env? (and maybe if the CI env var isn't set?)

Reply:
Also there aren't any tests on this endpoint / logic. I know this is an internal API, but it still feels like we should test that we don't break this by accident. Though CI would probably fail if this broke, it would be much better if we had more insight into what broke to cause the failure.

Reply:
> Could we log a warning if this endpoint is used in production env? (and maybe if the CI env var isn't set?)

We are running the FTR suites against production cloud, so I'm not sure we should be logging a warning? At least we should not check the CI var env?

> Also there aren't any tests on this endpoint / logic

Yea, I know. I can add basic unit test to check that 1/ the route is registered and 2/ calling the route properly calls the migrator. However I have no idea how to go further than that, as the API call is actually testing in the FTR. We can't really functional test a component used by the FTR framework itself, can we? Any insight on how to test that on a 'higher' level that unit is welcome.

Reply:
I added integration tests on the endpoint to ensure it properly calls `migrator.runMigrations` with the correct parameters

Reply:
> I know this is an internal API

Is Platform going to promote any naming convention to clearly distinguish public (with BWC guarantees and documentation) and internal (no BWC guarantees, implicit `use at your own risk` warning and _optional_ documentation) API endpoints? In Spaces/Security we started to define internal routes with `/internal/` prefix instead of `/api/` as an experiment that has been working well enough so far.

I'm a bit concerned that users will quickly discover this endpoint and may start using it in any imaginable context without realizing all the consequences leading to hard-to-handle/diagnose support cases....

Reply:
> Is Platform going to promote any naming convention to clearly distinguish public (with BWC guarantees and documentation) and internal

We haven't discussed it yet, but as I can see the current PR uses elasticsearch convention for internal API starting resource name with underscore. `_migrate`

Reply:
> internal API starting resource name with underscore. _migrate

I had an impression that what is prefixed with underscore denotes `action`, so that we can distinguish it from resource ID that can be easily part of URL as well. 

Moreover it seems we start public actions with underscore as well, e.g. `_find`: https://www.elastic.co/guide/en/kibana/master/saved-objects-api-find.html

Reply:
> Moreover it seems we start public actions with underscore as well, e.g. _find:

Hm... It's an experimental API. Well, yeah, we need a more obvious way to declare it.

Reply:
Anyone against `/internal/saved_objects/_migrate`?

Reply:
I don't want to introduce another convention without a broad discussion. as of me, `internal` could work, yes.

Reply:
> We are running the FTR suites against production cloud

By the way, does it mean that `env.packageInfo` (`dist`, `buildNum` etc.) in this case is completely indistinguishable from the one we have when Kibana is run in production outside of functional tests?

Reply:
`/internal/saved_objects/_migrate` kinda breaks the SO route prefix. This is very minor, but forces to use a `/` based router when declaring routes instead of a `/saved_objects` one. Would `/saved_objects/internal/_migrate` be acceptable, or do we really want the route to starts with `/internal` ?

Reply:
nevermind, did not understand the `internal` prefix was a replacement for the `api` one. Please forget my previous question